<a href="https://colab.research.google.com/github/micaeladepaolocabarcos/proyectooilandgas/blob/main/U6_Limpieza_y_An%C3%A1lisis_Exploratorio_de_un_Dataset_Real.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Práctica: Limpieza y Análisis Exploratorio de un Dataset Real**
### Producción de pozos de gas y petróleo no convencional (Vaca Muerta, Argentina)
**Fuente:** Secretaría de Energía de la Nación — [datos.gob.ar](https://www.datos.gob.ar/dataset/produccion-de-petroleo-y-gas-por-pozo)

https://colab.research.google.com/drive/1iTV2_7uZcpk6xjukBHCbeuPch4fdmB6g?usp=sharing

####**1. Inicialización y carga**

In [44]:
# Importo las librerías que voy a usar
import pandas as pd
import numpy as np

# Muestro todas las columnas cuando imprimo un DataFrame
pd.set_option('display.max_columns', 50)

# Muestro los números con dos decimales para que sea más fácil leerlos
pd.set_option('display.float_format', '{:,.2f}'.format)

In [45]:
# Conecto con Google Drive para no tener que subir el CSV en cada sesión (lo uso solo en Colab)
# from google.colab import drive
# drive.mount('/content/drive')

# Defino la ruta del archivo
RUTA_CSV = '/content/drive/MyDrive/Coderhouse/produccion-de-pozos-de-gas-y-petrleo-no-convencional.csv'

In [46]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [47]:
df_original = pd.read_csv('/content/produccion-de-pozos-de-gas-y-petrleo-no-convencional.csv', low_memory=False)

# Trabajo sobre una copia para conservar el original intacto
df = df_original.copy()

print('Filas:', df.shape[0])
print('Columnas:', df.shape[1])

Filas: 421022
Columnas: 40


In [48]:
# Miro las primeras filas para conocer la estructura
df.head()

,idempresa,anio,mes,idpozo,prod_pet,prod_gas,prod_agua,iny_agua,iny_gas,iny_co2,iny_otro,tef,vida_util,tipoextraccion,tipoestado,tipopozo,observaciones,fechaingreso,rectificado,habilitado,idusuario,empresa,sigla,formprod,profundidad,formacion,idareapermisoconcesion,areapermisoconcesion,idareayacimiento,areayacimiento,cuenca,provincia,coordenadax,coordenaday,tipo_de_recurso,proyecto,clasificacion,subclasificacion,sub_tipo_recurso,fecha_data
0,YSUR,2016,1,137601,52.10,490.49,6.38,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2016-02-17 10:50:46.929347,f,t,5,YSUR ENERGÍA ARGENTINA S.R.L.,APA.RN.EFO-125(d),LAJA,"3,785.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.84,-39.02,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2016-01-31
1,YSUR,2018,1,158251,258.12,395.19,35.34,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2018-02-10 08:37:14.717426,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-121(d),LAJA,"4,505.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.86,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
2,YSUR,2017,1,157357,879.59,"1,753.20",176.69,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2017-02-16 13:45:37.233373,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-313(d),LAJA,"3,713.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.82,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2017-01-31
3,YSUR,2018,1,157882,571.08,991.33,9.09,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2018-02-10 08:37:14.717426,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-301(d),LAJA,"3,800.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.82,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
4,YSUR,2016,1,134008,8.88,325.92,20.14,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2016-02-17 10:50:46.929347,f,t,5,YSUR ENERGÍA ARGENTINA S.R.L.,AEA.RN.EFO-92,LAJA,"3,832.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.89,-39.01,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2016-01-31


####**2. Diagnóstico del estado de los datos**

*2.1 Tipos de datos y memoria*

In [49]:
# Reviso el tipo de cada columna, la cantidad de valores no nulos y el uso de memoria
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421022 entries, 0 to 421021
Data columns (total 40 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idempresa               421022 non-null  object 
 1   anio                    421022 non-null  int64  
 2   mes                     421022 non-null  int64  
 3   idpozo                  421022 non-null  int64  
 4   prod_pet                421022 non-null  float64
 5   prod_gas                421022 non-null  float64
 6   prod_agua               421022 non-null  float64
 7   iny_agua                421022 non-null  float64
 8   iny_gas                 421022 non-null  float64
 9   iny_co2                 421022 non-null  float64
 10  iny_otro                421022 non-null  float64
 11  tef                     421022 non-null  float64
 12  vida_util               8952 non-null    float64
 13  tipoextraccion          420408 non-null  object 
 14  tipoestado          

**Observaciones:**
- El dataset tiene 421.022 filas y 40 columnas; cada fila es la producción de un pozo en un mes.
- Las fechas (`fecha_data`, `fechaingreso`) están cargadas como texto, no como fecha: hay que convertirlas para poder operar con ellas.
- `rectificado` y `habilitado` guardan valores lógicos como texto (`'t'`/`'f'`).
- Muchas columnas de texto repiten pocos valores (provincia, formación, tipo de pozo): convertirlas a `category` reduce memoria.
- `vida_util` y `observaciones` tienen muy pocos valores no nulos.

*2.2 Estadística descriptiva de las variables numéricas principales*

In [50]:
# Selecciono las variables de producción, profundidad y tiempo efectivo
variables_numericas = ['prod_pet', 'prod_gas', 'prod_agua', 'profundidad', 'tef']

# Calculo las estadísticas descriptivas
df[variables_numericas].describe()

,prod_pet,prod_gas,prod_agua,profundidad,tef
count,"421,022.00","421,022.00","421,022.00","421,022.00","421,022.00"
mean,333.70,629.71,188.77,"3,765.71",21.97
std,937.98,"1,628.89",625.46,"6,536.84",12.18
min,-0.00,-12.27,0.00,0.00,0.00
25%,0.00,8.23,0.00,"2,858.00",14.63
50%,7.79,103.64,12.65,"3,272.00",29.19
75%,122.31,449.96,86.48,"4,700.00",30.63
max,"26,593.26","40,086.80","34,792.95","378,939.00",79.34


**Observaciones:**
- Las producciones tienen medias muy por encima de las medianas (por ejemplo `prod_pet`: media 333,7 vs mediana 7,8), señal de distribuciones muy asimétricas: pocos pozos producen muchísimo y muchos producen poco o nada.
- Aparecen mínimos negativos en `prod_pet` y `prod_gas`, físicamente imposibles.
- `profundidad` tiene mínimo 0 y máximo 378.939 m, ambos imposibles.
- `tef` (días efectivos de producción en el mes) llega a 79,34, cuando un mes tiene como máximo 31 días.

*2.3 Valores nulos*

Primero uso el método pedido en la consigna (`.isnull().sum()`) y después lo paso a porcentaje para dimensionar el impacto de cada columna.

In [51]:
# Cuento los nulos por columna y me quedo solo con las que tienen alguno
nulos = df.isnull().sum()
nulos[nulos > 0]

,0
vida_util,412070
tipoextraccion,614
tipoestado,614
tipopozo,614
observaciones,397033
clasificacion,922
subclasificacion,922
sub_tipo_recurso,452


In [52]:
def resumen_nulos(dataframe):
    # Creo un diccionario vacío donde guardo el porcentaje de nulos de cada columna
    porcentajes = {}
    # Guardo la cantidad total de filas
    total_filas = len(dataframe)

    # Recorro cada columna del DataFrame
    for columna in dataframe.columns:
        # Cuento los nulos de la columna
        cantidad = dataframe[columna].isnull().sum()
        # Solo registro las columnas que tienen al menos un nulo
        if cantidad > 0:
            # Calculo el porcentaje y lo guardo en el diccionario
            porcentajes[columna] = round(cantidad / total_filas * 100, 2)

    # Convierto el diccionario en una Serie y la ordeno de mayor a menor
    return pd.Series(porcentajes, name='% nulos').sort_values(ascending=False)

resumen_nulos(df)

,% nulos
vida_util,97.87
observaciones,94.30
clasificacion,0.22
subclasificacion,0.22
tipopozo,0.15
tipoestado,0.15
tipoextraccion,0.15
sub_tipo_recurso,0.11


**Observaciones:**
- `vida_util` (97,87%) y `observaciones` (94,30%) están casi vacías: no tiene sentido imputarlas, conviene eliminarlas.
- El resto de las columnas con nulos tiene menos del 0,25% de faltantes, así que puedo tratarlas con imputación o eliminación puntual sin afectar el análisis.

*2.4 Duplicados*

Reviso dos tipos de duplicado: filas idénticas en todas sus columnas y filas repetidas según la clave lógica del dataset (un pozo reporta una sola vez por mes, así que `idpozo` + `anio` + `mes` debería ser única).

In [53]:
# Cuento las filas completamente duplicadas
print('Filas duplicadas (todas las columnas):', df.duplicated().sum())

# Cuento las filas duplicadas según la clave pozo-año-mes
clave = ['idpozo', 'anio', 'mes']
print('Filas duplicadas (clave pozo-año-mes):', df.duplicated(subset=clave).sum())

Filas duplicadas (todas las columnas): 0
Filas duplicadas (clave pozo-año-mes): 0


No hay duplicados de ningún tipo. Igualmente dejo el paso `drop_duplicates()` en la limpieza para que el flujo siga siendo válido si el archivo se actualiza (la fuente se actualiza mensualmente).

*2.5 Valores inválidos o inconsistentes*

Además de nulos y duplicados, busco valores que existen pero no tienen sentido físico.

In [54]:
# Cuento producciones negativas (no puede haber producción menor a cero)
for columna in ['prod_pet', 'prod_gas', 'prod_agua']:
    print(f'{columna} negativos:', (df[columna] < 0).sum())

# Cuento profundidades iguales a 0 (dato faltante cargado como cero) y mayores a 8000 m
print('profundidad = 0:', (df['profundidad'] == 0).sum())
print('profundidad > 8000:', (df['profundidad'] > 8000).sum())

# Cuento meses con más de 31 días de tiempo efectivo de producción
print('tef > 31 días:', (df['tef'] > 31).sum())

prod_pet negativos: 1
prod_gas negativos: 2
prod_agua negativos: 0
profundidad = 0: 5738
profundidad > 8000: 152
tef > 31 días: 15


In [55]:
# Miro qué valores aparecen en las profundidades mayores a 8000 m
df.loc[df['profundidad'] > 8000, 'profundidad'].value_counts().head()

,count
profundidad,
"378,939.00",122
"8,264.00",11
"8,239.00",10
"8,376.00",9


In [56]:
# Reviso las columnas de texto que en realidad representan valores lógicos (verdadero/falso)
for columna in ['rectificado', 'habilitado']:
    print(df[columna].value_counts(), '\n')

rectificado
f    420328
t       694
Name: count, dtype: int64 

habilitado
t    421022
Name: count, dtype: int64 



**Problemas detectados además de los nulos:**
- 3 producciones negativas (1 de petróleo y 2 de gas).
- 5.738 profundidades en 0, que en realidad son datos faltantes cargados como cero.
- 152 profundidades mayores a 8.000 m; la mayoría tiene exactamente 378.939 m, claramente un error de carga.
- 15 registros con `tef` mayor a 31 días.
- `habilitado` vale `'t'` en todas las filas, así que no aporta información.

####**3. Limpieza rigurosa**

Aplico las transformaciones en un orden lógico: primero elimino lo que no aporta información, después corrijo los tipos (para que los cálculos posteriores sean confiables), luego trato nulos y valores inválidos y por último, verifico duplicados y reseteo el índice.

*3.1 Eliminación de columnas sin información útil*

In [57]:
# Guardo la cantidad de filas y columnas antes de limpiar para comparar al final
filas_inicio = df.shape[0]
columnas_inicio = df.shape[1]

# Elimino vida_util, observaciones y habilitado
columnas_a_eliminar = ['vida_util', 'observaciones', 'habilitado']
df = df.drop(columns=columnas_a_eliminar)

print('Columnas eliminadas:', columnas_a_eliminar)
print('Columnas restantes:', df.shape[1])

Columnas eliminadas: ['vida_util', 'observaciones', 'habilitado']
Columnas restantes: 37


*3.2 Conversión de tipos de datos*

In [58]:
# Convierto las fechas de texto a datetime
df['fecha_data'] = pd.to_datetime(df['fecha_data'], format='%Y-%m-%d')
# fechaingreso tiene fracciones de segundo de largo variable, por eso uso el formato ISO8601
df['fechaingreso'] = pd.to_datetime(df['fechaingreso'], format='ISO8601')

# Convierto rectificado de 't'/'f' a booleano con un diccionario de equivalencias
df['rectificado'] = df['rectificado'].map({'t': True, 'f': False})

# Convierto a categoría las columnas de texto con pocos valores distintos que se repiten mucho
columnas_categoricas = ['idempresa', 'empresa', 'tipoextraccion', 'tipoestado', 'tipopozo',
                        'formprod', 'formacion', 'cuenca', 'provincia', 'tipo_de_recurso',
                        'proyecto', 'clasificacion', 'subclasificacion', 'sub_tipo_recurso']
for columna in columnas_categoricas:
    df[columna] = df[columna].astype('category')

# Verifico los tipos de las columnas que modifiqué
df[['fecha_data', 'fechaingreso', 'rectificado', 'tipopozo', 'formacion']].dtypes

,0
fecha_data,datetime64[ns]
fechaingreso,datetime64[ns]
rectificado,bool
tipopozo,category
formacion,category


In [59]:
# Comparo el uso de memoria antes y después de convertir los tipos
memoria_antes = df_original.memory_usage(deep=True).sum() / 1024**2
memoria_despues = df.memory_usage(deep=True).sum() / 1024**2
print(f'Memoria antes: {memoria_antes:,.1f} MB')
print(f'Memoria después: {memoria_despues:,.1f} MB')

Memoria antes: 635.6 MB
Memoria después: 182.0 MB


La conversión de tipos no es solo una cuestión de prolijidad: además de permitir operar con fechas (por ejemplo, obtener los días de cada mes en el paso 3.7), reduce la memoria del DataFrame de 635,6 MB a 182,0 MB (una reducción de más del 70%).

*3.3 Imputación por moda del pozo: `tipopozo`, `tipoextraccion`, `tipoestado`*

Estas tres columnas tienen nulos en las mismas filas. En lugar de imputar con la moda general del dataset (que asignaría "Gasífero" a cualquier pozo), uso la moda de cada pozo: si un pozo tiene otros meses informados, completo con el valor que más se repite en su propio historial.

In [60]:
# Verifico que los nulos de las tres columnas caen en las mismas filas
nulos_tipopozo = df['tipopozo'].isnull()
print('Coinciden con tipoextraccion:', (nulos_tipopozo == df['tipoextraccion'].isnull()).all())
print('Coinciden con tipoestado:', (nulos_tipopozo == df['tipoestado'].isnull()).all())

Coinciden con tipoextraccion: True
Coinciden con tipoestado: True


In [61]:
def moda_por_pozo(dataframe, columna):
    # Creo un diccionario donde cada pozo tiene su propio contador de valores
    contadores = {}

    # Recorro al mismo tiempo el id del pozo y el valor de la columna, fila por fila
    for pozo, valor in zip(dataframe['idpozo'], dataframe[columna]):
        # Salteo los valores nulos porque no aportan a la moda
        if pd.isnull(valor):
            continue
        # Si es la primera vez que veo este pozo, le creo un contador vacío
        if pozo not in contadores:
            contadores[pozo] = {}
        # Si es la primera vez que veo este valor en el pozo, lo inicializo en cero
        if valor not in contadores[pozo]:
            contadores[pozo][valor] = 0
        # Actualizo el contador
        contadores[pozo][valor] += 1

    # Creo el diccionario final con la moda de cada pozo
    modas = {}
    # Recorro cada pozo y su contador
    for pozo, conteo in contadores.items():
        # Me quedo con el valor que tiene el conteo más alto
        modas[pozo] = max(conteo, key=conteo.get)

    return modas

In [62]:
# Recorro las tres columnas y completo sus nulos con la moda del pozo
for columna in ['tipopozo', 'tipoextraccion', 'tipoestado']:
    nulos_antes = df[columna].isnull().sum()
    # Calculo la moda de cada pozo para esta columna
    modas = moda_por_pozo(df, columna)
    # Busco la moda correspondiente a cada fila según su pozo
    valores_imputados = df['idpozo'].map(modas)
    # Completo solo los nulos; paso a texto para poder combinar con la columna categórica
    df[columna] = df[columna].astype('object').fillna(valores_imputados).astype('category')
    nulos_despues = df[columna].isnull().sum()
    print(f'{columna}: {nulos_antes} nulos antes -> {nulos_despues} nulos después')

tipopozo: 614 nulos antes -> 4 nulos después
tipoextraccion: 614 nulos antes -> 4 nulos después
tipoestado: 614 nulos antes -> 4 nulos después


Con la moda del pozo recuperé 610 de las 614 filas con nulos. Las 4 restantes pertenecen a pozos que no tienen ningún mes con estos datos informados, así que no hay de dónde imputar: las elimino en el paso siguiente.

*3.4 Eliminación estratégica de filas con nulos no recuperables*

Las filas que quedaron sin `tipopozo` y las que no tienen `clasificacion`, `subclasificacion` o `sub_tipo_recurso` pertenecen a pozos que no tienen ningún mes informado en esas columnas, por lo que no hay de dónde imputar. Como representan una proporción mínima del total, las elimino.

In [63]:
# Elimino las filas que siguen teniendo nulos en columnas categóricas clave
columnas_clave = ['tipopozo', 'tipoextraccion', 'tipoestado', 'clasificacion', 'subclasificacion', 'sub_tipo_recurso']
filas_antes = len(df)
df = df.dropna(subset=columnas_clave)
filas_eliminadas = filas_antes - len(df)

print('Filas eliminadas:', filas_eliminadas)
print(f'Porcentaje del total: {filas_eliminadas / filas_antes * 100:.2f}%')

Filas eliminadas: 1378
Porcentaje del total: 0.33%


*3.5 Eliminación de producciones negativas*

In [64]:
# Me quedo solo con las filas donde las tres producciones son mayores o iguales a cero
filas_antes = len(df)
df = df[(df['prod_pet'] >= 0) & (df['prod_gas'] >= 0) & (df['prod_agua'] >= 0)]
print('Filas eliminadas por producción negativa:', filas_antes - len(df))

Filas eliminadas por producción negativa: 3


*3.6 Imputación de `profundidad` inválida con la mediana por formación*

Trato como dato faltante tanto las profundidades en 0 (el pozo no puede tener profundidad nula) como las mayores a 8.000 m: el valor que más se repite ahí es 378.939 m, físicamente imposible (los pozos de Vaca Muerta rondan los 2.500 a 4.000 m de profundidad vertical), lo que indica un error de carga.

Como la profundidad de un pozo depende sobre todo de la formación geológica a la que apunta, imputo con la mediana de su formación. Elijo la mediana en lugar de la media porque es resistente a valores extremos.

In [65]:
# Reemplazo los valores inválidos por NaN para tratarlos como faltantes
df['profundidad'] = df['profundidad'].astype('float64')
invalidas = (df['profundidad'] == 0) | (df['profundidad'] > 8000)
df.loc[invalidas, 'profundidad'] = np.nan
print('Profundidades marcadas como faltantes:', df['profundidad'].isnull().sum())

Profundidades marcadas como faltantes: 5890


In [66]:
def mediana_por_formacion(dataframe):
    # Creo un diccionario donde guardo la lista de profundidades válidas de cada formación
    profundidades = {}

    # Recorro al mismo tiempo la formación y la profundidad de cada fila
    for formacion, profundidad in zip(dataframe['formacion'], dataframe['profundidad']):
        # Salteo las profundidades faltantes
        if pd.isnull(profundidad):
            continue
        # Si es la primera vez que veo la formación, le creo una lista vacía
        if formacion not in profundidades:
            profundidades[formacion] = []
        # Agrego la profundidad a la lista de su formación
        profundidades[formacion].append(profundidad)

    # Calculo la mediana de cada formación
    medianas = {}
    for formacion, lista in profundidades.items():
        medianas[formacion] = np.median(lista)

    return medianas

# Calculo las medianas y completo los faltantes según la formación de cada fila
medianas = mediana_por_formacion(df)
df['profundidad'] = df['profundidad'].fillna(df['formacion'].astype('object').map(medianas))
print('Profundidades faltantes después de imputar:', df['profundidad'].isnull().sum())

Profundidades faltantes después de imputar: 0


*3.7 Corrección del tiempo efectivo (`tef`)*

`tef` son los días del mes en que el pozo estuvo produciendo, así que no puede superar la cantidad de días de ese mes. Recorto los valores excedidos al máximo posible usando la fecha del registro.

In [67]:
# Obtengo la cantidad de días de cada mes a partir de la fecha (ya convertida a datetime)
dias_del_mes = df['fecha_data'].dt.days_in_month
print('Registros con tef mayor a los días del mes:', (df['tef'] > dias_del_mes).sum())

# Recorto tef al máximo de días del mes
df['tef'] = df['tef'].clip(upper=dias_del_mes)
print('Después de corregir:', (df['tef'] > dias_del_mes).sum())

Registros con tef mayor a los días del mes: 71
Después de corregir: 0


*3.8 Duplicados*

In [68]:
# Elimino duplicados exactos y por clave (en este dataset no hay, pero dejo el paso para que el flujo sea reproducible si el archivo se actualiza)
df = df.drop_duplicates()
df = df.drop_duplicates(subset=['idpozo', 'anio', 'mes'])
print('Duplicados por clave pozo-año-mes:', df.duplicated(subset=['idpozo', 'anio', 'mes']).sum())

Duplicados por clave pozo-año-mes: 0


*3.9 Reseteo del índice y verificación final*

In [69]:
# Reseteo el índice porque al eliminar filas quedaron huecos en la numeración
# Uso drop=True para no guardar el índice viejo como una columna nueva
df = df.reset_index(drop=True)
df.index

RangeIndex(start=0, stop=419641, step=1)

In [70]:
# Verifico que las columnas numéricas sean realmente numéricas antes de hacer cálculos
for columna in ['prod_pet', 'prod_gas', 'prod_agua', 'profundidad', 'tef']:
    print(f'{columna}: {df[columna].dtype} -> numérica: {pd.api.types.is_numeric_dtype(df[columna])}')

prod_pet: float64 -> numérica: True
prod_gas: float64 -> numérica: True
prod_agua: float64 -> numérica: True
profundidad: float64 -> numérica: True
tef: float64 -> numérica: True


In [71]:
# Comparo el tamaño del dataset antes y después de la limpieza
print(f'Filas: {filas_inicio:,} -> {df.shape[0]:,} (se eliminaron {filas_inicio - df.shape[0]:,}, un {(filas_inicio - df.shape[0]) / filas_inicio * 100:.2f}%)')
print(f'Columnas: {columnas_inicio} -> {df.shape[1]}')

# Confirmo que no quedan nulos
print('Nulos totales restantes:', df.isnull().sum().sum())

Filas: 421,022 -> 419,641 (se eliminaron 1,381, un 0.33%)
Columnas: 40 -> 37
Nulos totales restantes: 0


####**4. Agregaciones de negocio**

**Unidades:** `prod_pet` y `prod_agua` están en m³; `prod_gas` en miles de m³.

*Pregunta 1: ¿Cómo evolucionó la producción anual total de petróleo y gas no convencional?*

In [72]:
# Agrupo por año y sumo la producción de petróleo y gas
produccion_anual = df.groupby('anio')[['prod_pet', 'prod_gas']].sum()

# Agrego la cantidad de pozos distintos que reportaron en cada año
produccion_anual['pozos_activos'] = df.groupby('anio')['idpozo'].nunique()
produccion_anual

,prod_pet,prod_gas,pozos_activos
anio,,,
2006,"15,067.28","425,467.59",199
2007,"20,821.97","414,711.99",207
2008,"39,629.87","699,831.65",250
2009,"65,410.50","1,009,414.97",146
2010,"66,063.40","1,230,472.88",195
2011,"144,142.20","1,833,805.90",244
2012,"303,274.99","2,391,052.62",340
2013,"555,595.45","2,916,657.69",530
2014,"1,121,582.49","4,942,706.07",876


**Interpretación:**
- La producción no convencional crece de forma sostenida: el petróleo pasa de unos 15 mil m³ en 2006 a casi 29,4 millones de m³ en 2025, y la cantidad de pozos activos se multiplica por más de 20.
- Durante la primera década el gas es el producto principal en volumen; desde 2021 el petróleo acelera fuertemente y en 2025 prácticamente alcanza al gas en volumen (recordar que el gas está en miles de m³).
- 2020 es el único año en que el gas cae (efecto pandemia), mientras que el petróleo siguió creciendo.
- 2026 es un año incompleto (los datos llegan hasta julio), por eso su total es menor: no debe leerse como una caída.

*Pregunta 2: ¿Qué formaciones rinden más por pozo?*

Calculo la producción mensual promedio por pozo en cada formación, separando pozos petrolíferos (petróleo) y gasíferos (gas) y considerando solo los meses en extracción efectiva para no promediar meses en que el pozo estuvo parado. Además, dejo afuera las formaciones con menos de 10 pozos, porque un promedio calculado sobre 1 o 2 pozos no es representativo.

In [73]:
# Me quedo con los meses en que el pozo estuvo produciendo
en_produccion = df[df['tipoestado'] == 'Extracción Efectiva']

# Pozos petrolíferos: promedio y cantidad de pozos por formación
petroliferos = en_produccion[en_produccion['tipopozo'] == 'Petrolífero']
rendimiento_pet = petroliferos.groupby('formacion', observed=True).agg(
    prod_pet_promedio=('prod_pet', 'mean'),
    cantidad_pozos=('idpozo', 'nunique'))
# Descarto las formaciones con menos de 10 pozos porque su promedio no es representativo
rendimiento_pet = rendimiento_pet[rendimiento_pet['cantidad_pozos'] >= 10]
rendimiento_pet.sort_values('prod_pet_promedio', ascending=False)

,prod_pet_promedio,cantidad_pozos
formacion,,
vaca muerta,913.50,2513
precuyo,120.62,10
lajas,117.84,29
agrio,110.68,14
quintuco,36.35,37
los molles,4.61,12


In [74]:
# Pozos gasíferos: promedio y cantidad de pozos por formación
gasiferos = en_produccion[en_produccion['tipopozo'] == 'Gasífero']
rendimiento_gas = gasiferos.groupby('formacion', observed=True).agg(
    prod_gas_promedio=('prod_gas', 'mean'),
    cantidad_pozos=('idpozo', 'nunique'))
# Aplico el mismo criterio de al menos 10 pozos por formación
rendimiento_gas = rendimiento_gas[rendimiento_gas['cantidad_pozos'] >= 10]
rendimiento_gas.sort_values('prod_gas_promedio', ascending=False).head(6)

,prod_gas_promedio,cantidad_pozos
formacion,,
vaca muerta,"3,623.59",844
agrio,"2,213.06",55
punta rosada,"1,512.93",110
mulichinco,923.58,241
magallanes,691.59,185
lajas,532.58,581


**Interpretación:**
- Vaca Muerta domina en ambos tipos de pozo: un pozo petrolífero produce en promedio unos 914 m³ de petróleo por mes, más de 7 veces lo que rinde la siguiente formación y concentra la gran mayoría de los pozos petrolíferos (2.513 pozos).
- En gas la diferencia es menor: Vaca Muerta lidera (unos 3.624 miles de m³/mes), seguida por Agrio y Punta Rosada. Lajas tiene muchos pozos gasíferos (581) pero con un rendimiento promedio mucho más bajo, consistente con ser un reservorio tight.
- Esto refuerza la decisión de analizar pozos petrolíferos y gasíferos por separado.

*Pregunta 3: ¿Qué empresas concentran la producción de petróleo?*

In [75]:
# Sumo el petróleo producido por cada empresa en todo el período
petroleo_por_empresa = df.groupby('empresa', observed=True)['prod_pet'].sum().sort_values(ascending=False)

# Calculo qué porcentaje del total representa cada empresa
participacion = (petroleo_por_empresa / petroleo_por_empresa.sum() * 100).round(2)

top_empresas = pd.DataFrame({'prod_pet_total': petroleo_por_empresa, 'participacion_%': participacion})
top_empresas.head(10)

,prod_pet_total,participacion_%
empresa,,
YPF S.A.,"81,798,102.29",58.24
VISTA ENERGY ARGENTINA SAU,"13,996,445.28",9.96
SHELL ARGENTINA S.A.,"9,797,705.78",6.98
PAN AMERICAN ENERGY SL,"7,505,730.73",5.34
PLUSPETROL S.A.,"4,922,509.86",3.50
TECPETROL S.A.,"4,188,210.55",2.98
VISTA OIL & GAS ARGENTINA SAU,"2,353,554.18",1.68
CHEVRON ARGENTINA S.R.L.,"2,213,027.33",1.58
EXXONMOBIL EXPLORATION ARGENTINA S.R.L.,"2,060,061.54",1.47


**Interpretación:**
- La producción de petróleo no convencional está muy concentrada: YPF sola aporta el 58,2% del total histórico y las 5 primeras empresas superan el 84%.
- Vista aparece con dos razones sociales distintas ("VISTA ENERGY ARGENTINA SAU" y "VISTA OIL & GAS ARGENTINA SAU"), probablemente por un cambio de denominación. Para un análisis por grupo empresario convendría unificarlas; acá las dejo como figuran en la fuente.

*Pregunta 4: ¿Cuántos pozos hay por provincia y de qué tipo de recurso (shale o tight)?*

In [76]:
# Cuento los pozos distintos por provincia y sub tipo de recurso
pozos_por_provincia = df.groupby(['provincia', 'sub_tipo_recurso'], observed=True)['idpozo'].nunique()

# Paso el sub tipo de recurso a columnas para leerlo como tabla cruzada
pozos_por_provincia = pozos_por_provincia.unstack(fill_value=0)
pozos_por_provincia['total'] = pozos_por_provincia.sum(axis=1)
pozos_por_provincia.sort_values('total', ascending=False)

sub_tipo_recurso,SHALE,TIGHT,total
provincia,,,
Neuquén,3420,1158,4578
Rio Negro,9,236,245
Santa Cruz,5,209,214
Mendoza,18,15,33
Chubut,3,1,4
Salta,0,3,3


**Interpretación:**
- Neuquén concentra el 90% de los pozos no convencionales del país y es la única provincia donde predomina el shale (3.420 pozos, casi todos de Vaca Muerta).
- En Río Negro y Santa Cruz el desarrollo es principalmente tight.
- Mendoza, Chubut y Salta tienen una presencia todavía marginal.

####**5. Conclusiones**

**Sobre la limpieza:**
- Partí de 421.022 filas y 40 columnas y terminé con 419.641 filas y 37 columnas, sin nulos. Solo eliminé el 0,33% de las filas: la mayoría de los problemas se resolvieron imputando en lugar de borrar.
- Las decisiones de imputación usan conocimiento del dominio: moda del propio pozo para las variables categóricas y mediana de la formación geológica para la profundidad.
- La conversión de tipos habilitó cálculos con fechas y redujo la memoria en más de un 70%.

**Sobre el negocio:**
- La producción no convencional argentina creció de forma sostenida y en los últimos años el crecimiento lo lidera el petróleo.
- Vaca Muerta es, por lejos, la formación más productiva por pozo, tanto en petróleo como en gas.
- La producción está muy concentrada, tanto geográficamente (Neuquén) como por operador (YPF).